packages/terra-time-input/src/TimeInput.jsx

```diff
@@ -210,59 +466,206 @@ class TimeInput extends React.Component {
+        if (this.state.hour) {
+          this.hourInput.textInput.setSelectionRange(this.state.hour.length, this.state.hour.length);
           event.preventDefault();
         }
       }
-    } else if (event.keyCode === keyCodes.ARROWLEFT || event.keyCode === keyCodes.DELETE || event.keyCode === keyCodes.BACKSPACE) {
-      if (type === inputType.MINUTE) {
-        // When the DELETE, BACK, or LEFT ARROW key is pressed and tf the cusor is at the left most position in the minute input, is empty or the cursor is before the value, move focus to the hour input when the left arrow is pressed.
-        if (this.minuteInput.textInput.selectionEnd === 0) {
-          this.hourInput.textInput.focus();
-          if (this.state.hour) {
-            this.hourInput.textInput.setSelectionRange(this.state.hour.length, this.state.hour.length);
-            event.preventDefault();
-          }
-        }
+    } else if (event.keyCode === keyCodes.ARROWRIGHT) {
+      // If the minute is empty or the cursor is after the value, move focus to the meridiem when
+      // the right arrow is pressed for 12 hour times.
+      if ((stateValue.length === 0 || stateValue.length === this.minuteInput.textInput.selectionEnd) && this.meridiemInput) {
+        this.meridiemInput.focus();
+        event.preventDefault();
       }
     }
 
     if (updateStateValue) {
-      this.handleValueChange(event, type, stateValue);
+      this.handleValueChange(event, inputType.MINUTE, stateValue, this.state.meridiem);
     }
   }
 
-  handleValueChange(event, type, timeValue) {
+  handleValueChange(event, type, timeValue, meridiem) {
     if (type === inputType.HOUR) {
-      this.setState({ hour: timeValue });
+      this.setState({
+        hour: timeValue,
+        meridiem,
+      });
     } else {
-      this.setState({ minute: timeValue });
+      this.setState({
+        minute: timeValue,
+      });
     }
 
-    if (this.props.onChange) {
-      let changedValue;
-      if (type === inputType.HOUR) {
-        changedValue = timeValue.concat(':', this.state.minute);
+    // Input values of length 1 indicate incomplete time, which means we cannot get a
+    // reliable time so onChange isn't triggered.
+    if (this.props.onChange && timeValue.length !== 1) {
+      const hour = type === inputType.HOUR ? timeValue : this.state.hour;
+      const minute = type === inputType.MINUTE ? timeValue : this.state.minute;
+
+      if (hour === '' && minute === '') {
+        this.props.onChange(event, '');
       } else {
-        changedValue = this.state.hour.concat(':', timeValue);
+        this.props.onChange(event, this.formatHour(hour, meridiem).concat(':', minute));
+      }
+    }
+  }
+
+  formatHour(hour, meridiem) {
+    if (!hour) {
+      return hour;
+    }
+
+    let tempHour = parseInt(hour, 10);
+
+    if (this.props.variant === FORMAT_12_HOUR) {
+      if (meridiem === this.postMeridiem && tempHour < 12) {
+        tempHour += 12;
```

Comment:
Should this be '+= 12' or '=12'

Reply:
Not here actually. For this variable, we want to add 12 to it if the time is not 12 and the meridiem is set to the postMeridiem, or set the tempHour to 0 if the time is 12 and the meridiem is anteMeridiem.